### PR TITLE
Add DCDD shock capture stabilization term to heat transfer

### PR DESCRIPTION
**Pull request by voferreira**
# Description of the problem

- Cross-wind flow leads to numerical instability in the heat transfer solver.
- Some of this instability was not being sufficiently dampened by the GGLS term we had before, especially at highly predominant convective problems (high Pe).

# Description of the solution

- Apply the Discontinuity-Capturing Directional Dissipation (DCDD) stabilization term by Tezduyar (2003).
- The term identifies crosswind regions and adds to them an artificial diffusivity, dampening the instabilities.

# How Has This Been Tested?

- Canonical circular advection-diffusion problem in steady state.
- Some tests needed to be updated.

# Documentation

- [x] To be improved

# Future changes

- Other terms exist that can be added, such as LSIC. For now, I believe this one will suffice.

### CODE REVIEW

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:77]:
```suggestion
      // In Tezduyar 2003, this is denoted r in equation (67)
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:86]: Which equation of the paper are you implementing here? Is it 79?

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:90]:
```suggestion
      // denoted s in equation (67).
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:95]:
```suggestion
      // Denoted (r.s)^2*ss in equation (70) of Tezduyar 2003.
      const Tensor<2, dim> k_corr =
        Utilities::fixed_power<2>(gradient_unit_vector * velocity_unit_vector) *
        outer_product(velocity_unit_vector, velocity_unit_vector);
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:96]:
```suggestion
      // Denoted rr in equation (70) of Tezduyar 2003.
      const Tensor<2, dim> gradient_unit_tensor =
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:98]:
```suggestion
      // Denoted rr - (r.s)^2*ss in equation (70) of Tezduyar 2003.
      const Tensor<2, dim> dcdd_factor = gradient_unit_tensor - k_corr;
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:218]: Can you add the same comment as the previous for the equations and terms?

**Review by hepap — CHANGES_REQUESTED** [review]
Nice addition!

I'm just not sure I understand how the equation 79 of the paper was converted for the temperature. Maybe you can explain it to me?

**Comment by voferreira** [comment]
@mivaia I have fixed the branch. You can restart from there.

File: applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.output
Comment (voferreira) [inline applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.output:91]: This file should not be modified

File: applications_tests/lethe-fluid-matrix-free/heat_transfer_high_pe_stab_gls.mpirun=2.output
Comment (voferreira) [inline applications_tests/lethe-fluid-matrix-free/heat_transfer_high_pe_stab_gls.mpirun=2.output:9]: Maybe double check this case. The error is twice as high.

File: applications_tests/lethe-fluid-matrix-free/heat_transfer_high_pe_stab_gls.output
Comment (voferreira) [inline applications_tests/lethe-fluid-matrix-free/heat_transfer_high_pe_stab_gls.output:31]: We need to double check this test.

File: applications_tests/lethe-fluid/high_pe_stab_gls.output
Comment (voferreira) [inline applications_tests/lethe-fluid/high_pe_stab_gls.output:31]: We will need to double-check the high Pe tests.

File: include/solvers/heat_transfer.h
Comment (voferreira) [inline include/solvers/heat_transfer.h:192]:
```suggestion
```

File: include/solvers/heat_transfer_scratch_data.h
Comment (voferreira) [inline include/solvers/heat_transfer_scratch_data.h:48]:
```suggestion
```

File: source/solvers/heat_transfer.cc
Comment (voferreira) [inline source/solvers/heat_transfer.cc:23]:
```suggestion
```

File: source/solvers/heat_transfer.cc
Comment (voferreira) [inline source/solvers/heat_transfer.cc:765]:
```suggestion
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (voferreira) [inline source/solvers/heat_transfer_assemblers.cc:10]:
```suggestion
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (voferreira) [inline source/solvers/heat_transfer_assemblers.cc:550]:
```suggestion
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (voferreira) [inline source/solvers/heat_transfer_assemblers.cc:848]:
```suggestion
```

File: include/solvers/heat_transfer.h
Comment (blaisb) [inline include/solvers/heat_transfer.h:196]: Can you add a return description here.

File: applications_tests/lethe-fluid/high_pe_stab_gls.output
Comment (blaisb) [reply applications_tests/lethe-fluid/high_pe_stab_gls.output:31]: Yeah, make sure that they converge. Here you seem to have affected the order of convergence significantly. It was already too high though, so might not be the worst thing, but still, double check that it is still consistant.

File: include/solvers/heat_transfer_scratch_data.h
Comment (blaisb) [inline include/solvers/heat_transfer_scratch_data.h:102]: not sure it's worth it to pass a reference to a double, just pass by copy instead.

File: source/solvers/heat_transfer.cc
Comment (blaisb) [inline source/solvers/heat_transfer.cc:776]: In some case it's present and in some case it'S previous. Call it solution_maximum and solution_minimum instead of previous then.

File: source/solvers/heat_transfer.cc
Comment (blaisb) [inline source/solvers/heat_transfer.cc:767]:
```suggestion
  double solution_maximum, solution_minimum;
```

File: source/solvers/heat_transfer.cc
Comment (blaisb) [inline source/solvers/heat_transfer.cc:764]: This function is super generic,  you could move it outside of the class and put it as a utility

File: source/solvers/heat_transfer.cc
Comment (blaisb) [inline source/solvers/heat_transfer.cc:780]: And this part the (1) should be an argument of the function with a default value of 1. .This way we expose that the function "clips" to a minimum value and we can change that value if need be,.

**Review by blaisb — CHANGES_REQUESTED** [review]
First batch of comments

**Review by mivaia — COMMENTED** [review]
Should be ready for a review. Adressed some of the comments that were already made.

File: source/solvers/heat_transfer_assemblers.cc
Comment (voferreira) [inline source/solvers/heat_transfer_assemblers.cc:285]:
```suggestion
          local_rhs(i) -=
```

File: source/solvers/heat_transfer.cc
Comment (voferreira) [reply source/solvers/heat_transfer.cc:764]: I prefer it here. Also, I feel like T_minimum being 1 always is not exactly a problem. But if you find it better as utilities so it is portable for other parts of the code, I don't mind. Also parsing the argument is not a big deal.

File: applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.output
Comment (voferreira) [reply applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.output:91]: That's true. But then, we could just remove the extra spaces.

File: include/solvers/heat_transfer.h
Comment (blaisb) [inline include/solvers/heat_transfer.h:200]: If you want you can put a default value of 1, but otherwise, I think this is a good version of the function

File: include/solvers/heat_transfer_scratch_data.h
Comment (blaisb) [inline include/solvers/heat_transfer_scratch_data.h:109]: I would add an assertion in the constructor to check if that value is >0 just to be defensive. You never know and this cost nothing

File: source/solvers/heat_transfer_assemblers.cc
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:73]: I think this could be more readable if done in a single line
```suggestion
      Tensor<1, dim> dcdd_temperature_gradient = is_steady(method) ?  scratch_data.temperature_gradients[q] : scratch_data.previous_temperature_gradients[0][q] ; 
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:219]: Same comment as above

**Review by blaisb — CHANGES_REQUESTED** [review]
1-2 few things. Good to go. How about the convergence of the high Péclet case, is it affected?

**Comment by blaisb** [comment]
@voferreira please avoid merging master into your branch, you should try to rebase only.

**Comment by voferreira** [comment]
@blaisb My bad. I will rebase from the next time on.

**Comment by blaisb** [comment]
is this ready to review or? The flags says ready to review but the tests do not pass.

**Comment by voferreira** [comment]
It is ready to be reviewed. There is only one test failing. It is a new test and it is just missing updating. I will update it when I arrive at the office.

**Comment by blaisb** [comment]
@voferreira What about my comment regarding convergence? I did not see any answer for the high Pe case.

File: applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.output
Comment (blaisb) [inline applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.output:10]: Why is the test case without VOF now? I don't understand, this used to be  VOF test case...

**Review by blaisb — COMMENTED** [review]
I'm good with the PR, but there are two points remaining
1- Why was the vof test changed to be without VOF now? If this test is without VOF now, the name of the test should be changed and the reason for this change should be explained in the PR ( I am unsure why such a change would be necessary)
2- What happens to the convergence of the high Pe case?

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:138]: Is this GLS or only SUPG? I'm always confused about the name!

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:223]:
```suggestion
      // Directionality tensor [rr - (r⋅s)^2*ss]
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (hepap) [inline source/solvers/heat_transfer_assemblers.cc:266]: See my other comment
```suggestion
          // Apply SUPG
```

**Comment by voferreira** [comment]
@blaisb Sorry for the delay about the comment on the high Pe, but I wanted to confirm with Mikael before. The test was not passing when: 1- we were using the previous solution for steady and transient, and 2- when we were testing the term without the velocity component. Now the test is not affected anymore. As for the VOF test, it was changed by mistake and I am fixing it just now.

**Comment by blaisb** [comment]
Good for me then, i'll wait for the other to approve before merging

**Review by PierreLaurentinCS — CHANGES_REQUESTED** [review]
Very fine addition, good job, just switch the indices in the assembly and we're good

File: source/solvers/heat_transfer_assemblers.cc
Comment (voferreira) [reply source/solvers/heat_transfer_assemblers.cc:74]: As we discussed, it is not. It is an adaptation of the term in Tezduyar 2003.

File: source/solvers/heat_transfer_assemblers.cc
Comment (voferreira) [reply source/solvers/heat_transfer_assemblers.cc:94]: As we discussed, this choice is debatable. We can experiment with other characteristic lengths, but I believe we have a good implementation for the time being. It is also coherent with SUPG here.

**Review by hepap — APPROVED** [review]

**Review by PierreLaurentinCS — APPROVED** [review]
Everything is fine, good job!

**Pull request merged**
